Adjust the calleeSavedRegs on top frame for LoongArch64/RISCV64

shushanhf opened the pull request:
Adjust the calleeSavedRegs on top frame for LoongArch64/RISCV64 to support the GSCookie.
```C
The frame layout:
  |                       |
  |-----------------------|
  |  incoming arguments   |
  +=======================+ <---- Caller's SP
  |  Varargs regs space   | // Only for varargs main functions; not used for LA64.
  |-----------------------|
  |    MonitorAcquired    | // 8 bytes; for synchronized methods
  |-----------------------|
  |        PSP slot       | // 8 bytes (omitted in NativeAOT ABI)
  |-----------------------|
  |Callee saved registers | // multiple of 8 bytes, not including FP/RA
  |-----------------------|
  |      Saved RA         | // 8 bytes
  |-----------------------|
  |      Saved FP         | // 8 bytes
  |-----------------------|
  |  possible GS cookie   |
  |-----------------------|
  | locals, temps, etc.   |
  |-----------------------|
  |  possible GS cookie   |
  |-----------------------|
  |   Outgoing arg space  | // multiple of 8 bytes; if required (i.e., #outsz != 0)
  |-----------------------| <---- Ambient SP
  |       |               |
  ~       | Stack grows   ~
  |       | downward      |
```

shushanhf commented:
@jakobbotsch  @clamp03 
Could you please review this PR?
Thanks

clamp03 commented:
@bartlomiejko Can you review and test this PR?
cc @dotnet/samsung

jakobbotsch commented:
How is the stack frame layout changed compared to previously? What was the layout before this change? Why didn't it support GS cookie?
Can you show some codegen diffs?

shushanhf commented:
> How is the stack frame layout changed compared to previously? What was the layout before this change? 

Only changed the calledSavedRegs, especially the FP/RA for LA64/RV64.  The layout liking this PR's first comment description.
Before this PR, the FP/RA at bottom of frame which under the GSCookie, the FP/RA can be overwritten.

> Why didn't it support GS cookie? Can you show some codegen diffs?

Before this PR, the FP/RA at bottom of frame which under the GSCookie, the FP/RA can be overwritten.

shushanhf commented on src/coreclr/jit/codegenloongarch64.cpp:
```diff
@@ -7660,46 +7531,50 @@ void CodeGen::instGen_MemoryBarrier(BarrierKind barrierKind)
- *      |  possible GS cookie   |
+ *      |      Saved RA         | // 8 bytes
  *      |-----------------------|
  *      |      Saved FP         | // 8 bytes
```
@jakobbotsch 
I just move the FP/RA to the top frame which always saved with calleeSavedRegs together.

shushanhf replied:
While the `MonitorAcquired` and `PSPSym` on the top of the frame.  The layout liking this PR's first comment description.

jakobbotsch replied:
"possible GS cookie" appears twice in this table. Should one be removed?

shushanhf replied:
> "possible GS cookie" appears twice in this table. Should one be removed?

The ARMARCH/XARCH also like this as the different condition to save it at different position within the lclvar.cpp.

shushanhf replied:
> > "possible GS cookie" appears twice in this table. Should one be removed?
> 
> The ARMARCH/XARCH also like this as the different condition to save it at different position within the lclvar.cpp.

The line 5488 within lclvars.cpp.

jakobbotsch commented on src/coreclr/jit/lclvars.cpp:
```diff
@@ -5793,8 +5822,8 @@ void Compiler::lvaFixVirtualFrameOffsets()
+        // For LoongArch64 and RISCV64, the RA is at fp+8.
+        lvaTable[lvaRetAddrVar].SetStackOffset(REGSIZE_BYTES);
     }
 #endif // !TARGET_LOONGARCH64
```
This can be unified with the above now it looks like.

shushanhf replied:
OK

jakobbotsch commented:
> Only changed the calledSavedRegs, especially the FP/RA for LA64/RV64. The layout liking this PR's first comment description [#100962 (comment)](https://github.com/dotnet/runtime/pull/100962#issue-2240209339) .
> Before this PR, the FP/RA at bottom of frame which under the GSCookie, the FP/RA can be overwritten.

Where does FP point to in your picture? Where did it point to before?

shushanhf commented:
> > Only changed the calledSavedRegs, especially the FP/RA for LA64/RV64. The layout liking this PR's first comment description [#100962 (comment)](https://github.com/dotnet/runtime/pull/100962#issue-2240209339) .
> > Before this PR, the FP/RA at bottom of frame which under the GSCookie, the FP/RA can be overwritten.
> 
> Where does FP point to in your picture? Where did it point to before?

The new FP is pointing to the the old FP saved slot.

> Can you show some codegen diffs?
```
/// this is the Old one.
G_M22522_IG01:              ;; offset=0x0000
  0xff74a50030  02FDC063  addi.d       sp, sp, -144
  0xff74a50034  29C00061  st.d         ra, sp, 0
  0xff74a50038  29C02076  st.d         fp, sp, 8
  0xff74a5003c  02C02076  addi.d       fp, sp, 8
  0xff74a50040  02C062CC  addi.d       t0, fp, 24 
  0xff74a50044  02801407  addi.w       a3, zero, 5
  0xff74a50048  29C02180  st.d         zero, t0, 8
  0xff74a5004c  29C00180  st.d         zero, t0, 0
  0xff74a50050  02FFFCE7  addi.d       a3, a3, -1
  0xff74a50054  02C0418C  addi.d       t0, t0, 16 
  0xff74a50058  5FFFF0E0  bne          a3, zero, 0xff74a50048
  0xff74a5005c  29C00180  st.d         zero, t0, 0
  0xff74a50060  29C202C4  st.d         a0, fp, 128
  0xff74a50064  29C1E2C5  st.d         a1, fp, 120
  0xff74a50068  29C1C2C6  st.d         a2, fp, 112
                        ;; size=60 bbWeight=1 PerfScore 0.00
G_M22522_IG02:              ;; offset=0x003C
```

```
/// This is the new ins.
G_M22522_IG01:              ;; offset=0x0000
  0xff77790030  02FDC063  addi.d       sp, sp, -144
  0xff77790034  29C20076  st.d         fp, sp, 128
  0xff77790038  29C22061  st.d         ra, sp, 136
  0xff7779003c  02C20076  addi.d       fp, sp, 128
  0xff77790040  02FE42CC  addi.d       t0, fp, -112
  0xff77790044  02801407  addi.w       a3, zero, 5
  0xff77790048  29C02180  st.d         zero, t0, 8
  0xff7779004c  29C00180  st.d         zero, t0, 0
  0xff77790050  02FFFCE7  addi.d       a3, a3, -1
  0xff77790054  02C0418C  addi.d       t0, t0, 16 
  0xff77790058  5FFFF0E0  bne          a3, zero, 0xff77790048  /// the upstream's disasm format had been updated.
  0xff7779005c  29C00180  st.d         zero, t0, 0
  0xff77790060  29FFE2C4  st.d         a0, fp, -8
  0xff77790064  29FFC2C5  st.d         a1, fp, -16
  0xff77790068  29FFA2C6  st.d         a2, fp, -24
                        ;; size=60 bbWeight=1 PerfScore 0.00
```

jakobbotsch commented:
If I understand correctly, the change here means you have to use negative offsets from FP to access locals more often. Does LA64/RV64 allow encoding as many negative offsets as it allows positive offsets?

shushanhf commented:
> *      |Callee saved registers | // not including FP/RA; multiple of 8 bytes

The GS Cookie is only used liking the localalloc. After the finished the frame allocation, when using the localalloc to alocate a new stack space, the new space maybe overwritten the new real size allocated while the GSCookie is used to recognize this case.

shushanhf commented:
> If I understand correctly, the change here means you have to use negative offsets from FP to access locals more often. Does LA64/RV64 allow encoding as many negative offsets as it allows positive offsets?

yes

jakobbotsch commented:
> The GS Cookie is only used liking the localalloc. After the finished the frame allocation, when using the localalloc to alocate a new stack space, the new space maybe overwritten the new real size allocated while the GSCookie is used to recognize this case.
> Before this PR which the FP/RA under GS cookie, the FP/RA maybe overwritten while the GS cookie may not recognize this.

GS cookie is not only used with `localloc`, but for any method with general unsafe structs. For example,
```csharp
public static int Test()
{
    S s = new S();
    return s.foo[32];
}

public struct S
{
    public fixed byte foo[64];
}
```
has GS cookie without `localloc`. For arm64 we allocate FP/LR at the bottom of the frame in this case, the locals above it, and the GS cookie above the locals:
```asm
; Assembly listing for method Program:Test():int (Tier0)
; Emitting BLENDED_CODE for generic ARM64 - Windows
; Tier0 code
; fp based frame
; partially interruptible
; invoked as altjit
; Final local variable assignments
;
;  V00 loc0         [V00    ] (  1,  1   )  struct (64) [fp+0x18]  do-not-enreg[SF] must-init ld-addr-op unsafe-buffer <Program+S>
;  V01 tmp0         [V01    ] (  1,  1   )     int  ->  [fp+0x14]  do-not-enreg[V] "GSCookie dummy"
;# V02 OutArgs      [V02    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V03 GsCookie     [V03    ] (  1,  1   )    long  ->  [fp+0x58]  do-not-enreg[X] addr-exposed "GSSecurityCookie"
;
; Lcl frame size = 80

G_M58954_IG01:  ;; offset=0x0000
            stp     fp, lr, [sp, #-0x60]!
            mov     fp, sp
```

For the case you describe with localloc we allocate FP/LR at the top on arm64. For example:
```csharp
public static int Foo()
{
    int x = 1234;
    byte* pb = stackalloc byte[50];
    pb[24] = (byte)x;
    return pb[24];
}
```
...
```asm
; Assembly listing for method Program:Foo():int (Tier0)
; Emitting BLENDED_CODE for generic ARM64 - Windows
; Tier0 code
; fp based frame
; partially interruptible
; invoked as altjit
; Final local variable assignments
;
;  V00 loc0         [V00    ] (  1,  1   )     int  ->  [fp-0x04]  do-not-enreg[] must-init
;# V01 OutArgs      [V01    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
;  V02 tmp1         [V02    ] (  1,  1   )    long  ->  [fp-0x10]  do-not-enreg[] "DUP instruction"
;  V03 GsCookie     [V03    ] (  1,  1   )    long  ->  [fp-0x18]  do-not-enreg[X] addr-exposed "GSSecurityCookie"
;
; Lcl frame size = 32

G_M53658_IG01:  ;; offset=0x0000
            sub     sp, sp, #48
            stp     fp, lr, [sp, #0x20]
            add     fp, sp, #32
            str     wzr, [fp, #-0x04]	// [V00 loc0]
```
What you're saying is that LA64 did not do this before. That is, on LA64/RV64 you would still allocate the FP/RA at the bottom in this case. Do I understand correctly? Why didn't it do that before, given that the logic follows the general ARM64 logic?

This PR changes LA64/RV64 to always place FP/RA at the top. That doesn't result in any CQ regressions because LA/RISCV64 can encode negative offsets in addressing modes in equal manner to positive offsets, which is not true for arm64.
However, it makes OSR a bit different. For methods with patchpoints we always put FP/LR at the bottom on arm64, while this change ends up with FP/RA at the top. So that's why the general OSR paths are affected a bit.

Do I understand correctly?

jakobbotsch commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5935,6 +5941,12 @@ void Compiler::generatePatchpointInfo()
+    // NOTE:
+    //    The callee saved registers include the FP/RA here!!!
+    //    We save the first callee saved register's offset, which is the FP's offset.
+    patchpointInfo->SetCalleeSaveRegisters(codeGen->genSPtoFPdelta());
```
`CalleeSaveRegisters` is supposed to be a register mask. This PR starts using it for something else. Why is it necessary to save this offset in this way?
This kind of thing makes long-term maintenance hard for us.

shushanhf replied:
> `CalleeSaveRegisters` is supposed to be a register mask. This PR starts using it for something else. Why is it necessary to save this offset in this way? This kind of thing makes long-term maintenance hard for us.

As I don't want to add new interface.
If needed, I can add a new one.

shushanhf commented:
> > The GS Cookie is only used liking the localalloc. After the finished the frame allocation, when using the localalloc to alocate a new stack space, the new space maybe overwritten the new real size allocated while the GSCookie is used to recognize this case.
> > Before this PR which the FP/RA under GS cookie, the FP/RA maybe overwritten while the GS cookie may not recognize this.
> 
> GS cookie is not only used with `localloc`, but for any method with general unsafe structs. For example,

Yes, I just give an example above while the description mightbe not exact my mean.

> 
> ```cs
> public static int Test()
> {
>     S s = new S();
>     return s.foo[32];
> }
> 
> public struct S
> {
>     public fixed byte foo[64];
> }
> ```
> 
> has GS cookie without `localloc`. For arm64 we allocate FP/LR at the bottom of the frame in this case, the locals above it, and the GS cookie above the locals:
> 
> ```assembly
> ; Assembly listing for method Program:Test():int (Tier0)
> ; Emitting BLENDED_CODE for generic ARM64 - Windows
> ; Tier0 code
> ; fp based frame
> ; partially interruptible
> ; invoked as altjit
> ; Final local variable assignments
> ;
> ;  V00 loc0         [V00    ] (  1,  1   )  struct (64) [fp+0x18]  do-not-enreg[SF] must-init ld-addr-op unsafe-buffer <Program+S>
> ;  V01 tmp0         [V01    ] (  1,  1   )     int  ->  [fp+0x14]  do-not-enreg[V] "GSCookie dummy"
> ;# V02 OutArgs      [V02    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
> ;  V03 GsCookie     [V03    ] (  1,  1   )    long  ->  [fp+0x58]  do-not-enreg[X] addr-exposed "GSSecurityCookie"
> ;
> ; Lcl frame size = 80
> 
> G_M58954_IG01:  ;; offset=0x0000
>             stp     fp, lr, [sp, #-0x60]!
>             mov     fp, sp
> ```
> 
> For the case you describe with localloc we allocate FP/LR at the top on arm64. For example:
> 
> ```cs
> public static int Foo()
> {
>     int x = 1234;
>     byte* pb = stackalloc byte[50];
>     pb[24] = (byte)x;
>     return pb[24];
> }
> ```
> 
> ...
> 
> ```assembly
> ; Assembly listing for method Program:Foo():int (Tier0)
> ; Emitting BLENDED_CODE for generic ARM64 - Windows
> ; Tier0 code
> ; fp based frame
> ; partially interruptible
> ; invoked as altjit
> ; Final local variable assignments
> ;
> ;  V00 loc0         [V00    ] (  1,  1   )     int  ->  [fp-0x04]  do-not-enreg[] must-init
> ;# V01 OutArgs      [V01    ] (  1,  1   )  struct ( 0) [sp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
> ;  V02 tmp1         [V02    ] (  1,  1   )    long  ->  [fp-0x10]  do-not-enreg[] "DUP instruction"
> ;  V03 GsCookie     [V03    ] (  1,  1   )    long  ->  [fp-0x18]  do-not-enreg[X] addr-exposed "GSSecurityCookie"
> ;
> ; Lcl frame size = 32
> 
> G_M53658_IG01:  ;; offset=0x0000
>             sub     sp, sp, #48
>             stp     fp, lr, [sp, #0x20]
>             add     fp, sp, #32
>             str     wzr, [fp, #-0x04]	// [V00 loc0]
> ```
> 
> What you're saying is that LA64 did not do this before. That is, on LA64/RV64 you would still allocate the FP/RA at the bottom in this case. Do I understand correctly? 

Yes, you are right.

> Why didn't it do that before, given that the logic follows the general ARM64 logic?

This is an error while not test the GScookie completely.
 
> This PR changes LA64/RV64 to always place FP/RA at the top. That doesn't result in any CQ regressions because LA/RISCV64 can encode negative offsets in addressing modes in equal manner to positive offsets, which is not true for arm64. However, it makes OSR a bit different. For methods with patchpoints we always put FP/LR at the bottom on arm64, while this change ends up with FP/RA at the top. So that's why the general OSR paths are affected a bit.

Yes. you are right.

shushanhf commented:
> This PR changes LA64/RV64 to always place FP/RA at the top. That doesn't result in any CQ regressions because LA/RISCV64 can encode negative offsets in addressing modes in equal manner to positive offsets, which is not true for arm64.

BTW, I'm curious about the arm64 cannot encode negative offsets. Is the ARM64's manual required? I didn't find it. why?

jakobbotsch commented:
Thank you. So to sum up:
- Today, arm64 normally puts FP/LR at the bottom of the frame, below locals. This allows addressing locals with positive offsets which are more readily encodable on arm64.
- LA64/RISCV64 followed the same model in their bring-up, putting FP/RA at the bottom, below locals. Unlike arm64 there is no code quality benefit here as addressing modes can use negative offsets to the same extent as positive ones.
- When the function has `localloc` ARM64 switches to a frame layout that has FP/LR at the top, above locals, with a GS cookie between locals/localloc'd memory and FP/LR. This is done for security reasons. The ARM64 backend thus has to support multiple possible frame layouts.
- LA64/RISCV64 never implemented any other frame layouts, so in the above situation they keep FP/RA at the bottom, with no GS cookie between localloc'd memory and FP/RA. Thus FP/RA is not protected by the stack cookie in these situations.
- To fix the problem this PR starts always putting FP/RA at the top on these frames. Since there is no code quality benefit to keeping FP at the bottom this is a simpler approach than supporting multiple possible frame layouts.
- Unfortunately OSR patchpoint information does not encode enough information to properly transition out of the stack frames with FP/RA at the top, so the PR has to either add information to the patchpoint info or repurpose an existing field to communicate new information across.

Does that sum up the situation of this PR? I would still like to get @BruceForstall's input on this, but to me it seems reasonable to make this change. The problem/fix around the OSR issue is a little unsatisfying, perhaps @AndyAyersMS has some thoughts on this?

jakobbotsch commented:
> BTW, I'm curious about the arm64 cannot encode negative offsets. Is the ARM64's manual required? I didn't find it. why?

arm64 can encode negative offsets, but it has a larger range of encodable positive offsets. See the "pimm" encoding here: https://developer.arm.com/documentation/dui0802/b/A64-Data-Transfer-Instructions/LDR--immediate-

shushanhf commented:
> Thank you. So to sum up:
> 
> * Today, arm64 normally puts FP/LR at the bottom of the frame, below locals. This allows addressing locals with positive offsets which are more readily encodable on arm64.
> * LA64/RISCV64 followed the same model in their bring-up, putting FP/RA at the bottom, below locals. Unlike arm64 there is no code quality benefit here as addressing modes can use negative offsets to the same extent as positive ones.
> * When the function has `localloc` ARM64 switches to a frame layout that has FP/LR at the top, above locals, with a GS cookie between locals/localloc'd memory and FP/LR. This is done for security reasons. The ARM64 backend thus has to support multiple possible frame layouts.
> * LA64/RISCV64 never implemented any other frame layouts, so in the above situation they keep FP/RA at the bottom, with no GS cookie between localloc'd memory and FP/RA. Thus FP/RA is not protected by the stack cookie in these situations.
> * To fix the problem this PR starts always putting FP/RA at the top on these frames. Since there is no code quality benefit to keeping FP at the bottom this is a simpler approach than supporting multiple possible frame layouts.
> * Unfortunately OSR patchpoint information does not encode enough information to properly transition out of the stack frames with FP/RA at the top, so the PR has to either add information to the patchpoint info or repurpose an existing field to communicate new information across.
> 
> Does that sum up the situation of this PR? I would still like to get @BruceForstall's input on this, but to me it seems reasonable to make this change. The problem/fix around the OSR issue is a little unsatisfying, perhaps @AndyAyersMS has some thoughts on this?

Thanks very much!

shushanhf commented:
> > BTW, I'm curious about the arm64 cannot encode negative offsets. Is the ARM64's manual required? I didn't find it. why?
> 
> arm64 can encode negative offsets, but it has a larger range of encodable positive offsets. See the "pimm" encoding here: https://developer.arm.com/documentation/dui0802/b/A64-Data-Transfer-Instructions/LDR--immediate-

OK, thanks! that's just for large offset range.
I understand.
Thanks

AndyAyersMS commented:
> The problem/fix around the OSR issue is a little unsatisfying, perhaps @AndyAyersMS has some thoughts on this?

We don't support OSR for localloc methods so perhaps that's why I never added support for those frame types for arm64.

If you need more information, please add another field to the patchpoint info. You will also need to bump the JIT GUID as this information is serialized out to SPMI.

shushanhf commented:
> > The problem/fix around the OSR issue is a little unsatisfying, perhaps @AndyAyersMS has some thoughts on this?
> 
> We don't support OSR for localloc methods so perhaps that's why I never added support for those frame types for arm64.
> 
> If you need more information, please add another field to the patchpoint info. You will also need to bump the JIT GUID as this information is serialized out to SPMI.

OK, Thanks very much!

shushanhf commented:
> > The problem/fix around the OSR issue is a little unsatisfying, perhaps @AndyAyersMS has some thoughts on this?
> 
> We don't support OSR for localloc methods so perhaps that's why I never added support for those frame types for arm64.
> 
> If you need more information, please add another field to the patchpoint info. You will also need to bump the JIT GUID as this information is serialized out to SPMI.

Hi, @AndyAyersMS 
Is the JIT GUID here `constexpr GUID JITEEVersionIdentifier = {...}` within the file `src/coreclr/inc/jiteeversionguid.h`

AndyAyersMS commented:
Yes, that's the GUID.

shushanhf commented:
> Yes, that's the GUID.

Thanks

jakobbotsch commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -5834,12 +5834,18 @@ void Compiler::generatePatchpointInfo()
     //
     const int totalFrameSize = codeGen->genTotalFrameSize();
     const int offsetAdjust   = codeGen->genSPtoFPdelta() - totalFrameSize;
+#elif defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
+    // SP is not manipulated by calls so no frame size adjustment needed.
+    // Local Offsets don't need adjusting for LoongArch64.
+    //
+    const int totalFrameSize = codeGen->genTotalFrameSize();
+    const int offsetAdjust   = 0;
```
I don't understand why we need to add a field to the patchpoint information. Can't we just change the `offsetAdjust` here such that the offsets are encoded relative to the bottom of the frame like before this change? That should give us all the necessary information to compute the offsets into the tier 0 frame.

shushanhf replied:
The patchpoint information should save the whole tier0 frame’s size and the tier0's FP relative to SP.
They are all needed by tier1.

jakobbotsch replied:
What is tier0 FP relative to SP needed for in the tier1 frame?

shushanhf replied:
> What is tier0 FP relative to SP needed for in the tier1 frame?

The locals which are shared with tier1 frame, their tier0's offsets should be updated to relative tier1's FP.

shushanhf replied:
> What I am suggesting here is to encode their offsets in the patch point information in tier-0 so that this is unnecessary. Then the field isn't needed in tier-1 anymore.

What you mean is `offsetAdjust = codeGen->genSPtoFPdelta()`, and then avoid adding the `SetCalleeSaveRegistersOffset()`

jakobbotsch replied:
Yes, exactly.

shushanhf replied:
That's a good idea.
Thanks very much.

jakobbotsch commented on src/coreclr/jit/lclvars.cpp:
```diff
@@ -5671,23 +5671,52 @@ void Compiler::lvaFixVirtualFrameOffsets()
         delta += codeGen->genTotalFrameSize() - codeGen->genSPtoFPdelta();
     }
-#endif // TARGET_AMD64 || TARGET_ARM64 || TARGET_LOONGARCH64 || TARGET_RISCV64
+#elif defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
+    else
+    {
+        // FP is used.
+        delta += (compCalleeRegsPushed << 3);
+
+        if ((lvaMonAcquired != BAD_VAR_NUM) && !opts.IsOSR())
+        {
+            int offset = lvaTable[lvaMonAcquired].GetStackOffset() + delta;
+            lvaTable[lvaMonAcquired].SetStackOffset(offset);
+
+            if (lvaPSPSym != BAD_VAR_NUM)
+            {
+                int offset = lvaTable[lvaPSPSym].GetStackOffset() + delta;
+                lvaTable[lvaPSPSym].SetStackOffset(offset);
+                delta += TARGET_POINTER_SIZE;
+            }
+
+            delta += lvaLclSize(lvaMonAcquired);
+        }
+        else if (lvaPSPSym != BAD_VAR_NUM)
+        {
+            int offset = lvaTable[lvaPSPSym].GetStackOffset() + delta;
+            lvaTable[lvaPSPSym].SetStackOffset(offset);
+            delta += TARGET_POINTER_SIZE;
+        }
+
+        JITDUMP("--- delta bump %d for FP frame\n", delta);
+    }
+#endif // !TARGET_LOONGARCH64 || !TARGET_RISCV64
```
Why is this change necessary? Isn't `codeGen->genTotalFrameSize() - codeGen->genSPtoFPdelta()` the same value as this is computing?

jakobbotsch replied:
I guess the `SetStackOffset` is the important part, not the delta computation.

shushanhf replied:
> Why is this change necessary? Isn't `codeGen->genTotalFrameSize() - codeGen->genSPtoFPdelta()` the same value as this is computing?

Because the lvaMonAcquired and lvaPSPSym are saved on the top which their offset are positive and the offsert are greater than the calleeSavedRegs. These two offsets should be ajusted first and then will be computed by later `SetStackOffset` with other locals.

> I guess the `SetStackOffset` is the important part, not the delta computation.

Yes, this is important, but can not compute all info. liking the OutGoingArgs and RetAddr.

clamp03 commented:
> @bartlomiejko Can you review and test this PR? cc @dotnet/samsung

@bartlomiejko Did you assign someone in your team to check this PR on RISC-V?

bartlomiejko commented:
> > @bartlomiejko Can you review and test this PR? cc @dotnet/samsung
> 
> @bartlomiejko Did you assign someone in your team to check this PR on RISC-V?

Yes, we will check it.

shushanhf committed:
Adjust the calleeSavedRegs on top frame for LoongArch64/RISCV64
to support the GSCookie.
The frame layout:
   |                       |
   |-----------------------|
   |  incoming arguments   |
   +=======================+ <---- Caller's SP
   |  Varargs regs space   | // Only for varargs main functions; not used for LA64.
   |-----------------------|
   |    MonitorAcquired    | // 8 bytes; for synchronized methods
   |-----------------------|
   |        PSP slot       | // 8 bytes (omitted in NativeAOT ABI)
   |-----------------------|
   |Callee saved registers | // multiple of 8 bytes, not includting FP/RA
   |-----------------------|
   |      Saved RA         | // 8 bytes
   |-----------------------|
   |      Saved FP         | // 8 bytes
   |-----------------------|
   |  possible GS cookie   |
   |-----------------------|
   | locals, temps, etc.   |
   |-----------------------|
   |  possible GS cookie   |
   |-----------------------|
   |   Outgoing arg space  | // multiple of 8 bytes; if required (i.e., #outsz != 0)
   |-----------------------| <---- Ambient SP
   |       |               |
   ~       | Stack grows   ~
   |       | downward      |

shushanhf commented:
@jakobbotsch 
This error seems about the arm64's new ABI-style
https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-100962-merge-e5d222f8e30f445587/windows-arm64-3/1/console.5ac37b0c.log?helixlogtype=result
```
[10:19:40] ISSUE: <ASSERT> #15375 D:\a\_work\1\s\src\coreclr\jit\hwintrinsiccodegenarm64.cpp (1188) - Assertion failed 'argReg == argNode->GetRegNum()' in 'System.Buffers.Text.Base64:AdvSimdDecode(byref,byref,ulong,int,int,ulong,ulong)' during 'Generate code' (IL size 665; hash 0x2982fdff; FullOpts)
```
It's not about this PR.

shushanhf committed:
update code annotation.

jakobbotsch commented:
> @jakobbotsch This error seems about the arm64's new ABI-style https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-100962-merge-e5d222f8e30f445587/windows-arm64-3/1/console.5ac37b0c.log?helixlogtype=result
> 
> ```
> [10:19:40] ISSUE: <ASSERT> #15375 D:\a\_work\1\s\src\coreclr\jit\hwintrinsiccodegenarm64.cpp (1188) - Assertion failed 'argReg == argNode->GetRegNum()' in 'System.Buffers.Text.Base64:AdvSimdDecode(byref,byref,ulong,int,int,ulong,ulong)' during 'Generate code' (IL size 665; hash 0x2982fdff; FullOpts)
> ```
> 
> It's not about this PR.

#101070 tracks this issue. The issue seems related to allocation of consecutive registers on arm64, I don't think it's ABI related.

jakobbotsch commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -4118,7 +4118,37 @@ void CodeGen::genEnregisterOSRArgsAndLocals()
         genInstrWithConstant(ins_Load(lclTyp), size, varDsc->GetRegNum(), genFramePointerReg(), offset, initReg);
         *pInitRegZeroed = false;
-#endif // TARGET_ARM64 || TARGET_LOONGARCH64 || TARGET_RISCV64
+#elif defined(TARGET_LOONGARCH64) || defined(TARGET_RISCV64)
+
+        // Patchpoint offset is from top of Tier0 frame
+        //
+        // We need to determine the frame-pointer relative
+        // offset for this variable in the osr frame.
+        //
+        // First there is no need to ajust stkOffs
+        // as it relative to sp within Tier0 frame
+
+        // then add the OSR frame size
+        //
+        const int osrFrameSize = genTotalFrameSize();
+
+        // then subtract OSR SP-FP delta
+        //
+        const int osrSpToFpDelta = genSPtoFPdelta();
+
+        //                 | => tier0 top of frame relative
+        //                 |         + => osr bottom of frame (sp) relative
+        //                 |         |              - => osr fp relative
+        //                 |         |              |
+        const int offset = stkOffs + osrFrameSize - osrSpToFpDelta;
+
+        JITDUMP("---OSR--- V%02u (reg) Tier0 virtual offset %d OSR frame size %d OSR sp-fp "
+                "delta %d total offset %d (0x%x)\n",
+                varNum, stkOffs, osrFrameSize, osrSpToFpDelta, offset, offset);
+
+        genInstrWithConstant(ins_Load(lclTyp), size, varDsc->GetRegNum(), genFramePointerReg(), offset, initReg);
+        *pInitRegZeroed = false;
+#endif // TARGET_LOONGARCH64 || TARGET_RISCV64
```
Is any of this code necessary? Can we share LA64/RISCV64 OSR logic remain shared with ARM64 by changing
```
const int offsetAdjust = codeGen->genSPtoFPdelta();
```
to
```
const int offsetAdjust = codeGen->genSPtoFPdelta() - totalFrameSize;
```
in tier0?

shushanhf replied:
IMO, I think the ARM64's OSR is not so good here.
Maybe we should amend these code by a new PR ?

shushanhf replied:
> IMO, I think the ARM64's OSR is not so good here. Maybe we should amend all related OSR code by a new PR ?

For ARM64's OSR, first sub `info.compPatchpointInfo->TotalFrameSize()` on the line-6578 of `Compiler::lvaAssignVirtualFrameOffsetsToLocals`, then later add `info.compPatchpointInfo->TotalFrameSize()` on line-5719 of `Compiler::lvaFixVirtualFrameOffsets`.

shushanhf replied:
> I would prefer we keep ARM64/LA64/RISCV64 shared/uniform in this PR if we can.
> 
> I'd be fine with changing the offsets to be relative to something else in a new PR if that simplifies things. It would make sense for me if we made them relative to the bottom of the tier0 stack frame on all platforms, but it would be @AndyAyersMS's decision about the change.
> 
> Currently I'm not sure I understand what exactly the encoded offsets are relative to on the different platforms. Perhaps @AndyAyersMS can shed some light on this?

I agree with you.
It's best to discuss with @AndyAyersMS to get better design.

shushanhf replied:
@jakobbotsch 
I had push a new patch to uniform the ARM64 to LA64/RV64.

jakobbotsch replied:
The change to arm64 will need JIT-EE GUID update. I think x64 can also be changed such that offsets are always relative to the bottom of the frame for all targets. But I think it would be better to do this as a separate PR instead of doing it in this PR. For this PR it's best if you leave arm64/x64 unchanged and keep LA64/RISCV64 similar to arm64.

shushanhf replied:
Ok, thanks

AndyAyersMS replied:
> Currently I'm not sure I understand what exactly the encoded offsets are relative to on the different platforms. Perhaps @AndyAyersMS can shed some light on this?

At least for x64 and arm64, they are offsets are relative to the top of the Tier0 frame. 

How does (the idea of) changing this make things simpler?

shushanhf replied:
OK, thanks.
If needed, maybe it's better to discuss this in details by a new PR later.

am11 commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -4086,16 +4086,15 @@ void CodeGen::genEnregisterOSRArgsAndLocals()
-        // First add the Tier0 frame size
-        //
-        const int tier0FrameSize = compiler->info.compPatchpointInfo->TotalFrameSize();
+        // First there is no need to ajust stkOffs
+        // as it relative to sp within Tier0 frame
```
```suggestion
        // Note: there is no need to adjust stkOffs
        // as it is relative to sp within Tier0 frame.
```

shushanhf replied:
thanks

am11 commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -4086,16 +4086,15 @@ void CodeGen::genEnregisterOSRArgsAndLocals()
+        // First there is no need to ajust stkOffs
+        // as it relative to sp within Tier0 frame
 
         // then add the OSR frame size
```
```suggestion
        // Add the OSR frame size.
```

shushanhf replied:
thanks

am11 commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -4105,20 +4104,19 @@ void CodeGen::genEnregisterOSRArgsAndLocals()
         genInstrWithConstant(ins_Load(lclTyp), size, varDsc->GetRegNum(), genFramePointerReg(), offset, initReg);
         *pInitRegZeroed = false;
-#endif // TARGET_ARM64 || TARGET_LOONGARCH64 || TARGET_RISCV64
+#endif // TARGET_LOONGARCH64 || TARGET_RISCV64
```
```suggestion
#endif // TARGET_LOONGARCH64 || TARGET_RISCV64 || TARGET_ARM64
```

shushanhf replied:
thanks

tomeksowi commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -770,29 +770,26 @@ void CodeGen::genRestoreCalleeSavedRegistersHelp(regMaskTP regsToRestoreMask, in
  *      ~  alignment padding    ~ // To make the whole frame 16 byte aligned
  *      |-----------------------|
- *      |      Saved FP         | // 8 bytes
+ *      |Callee saved registers | // multiple of 8 bytes, not includting FP/RA
```
```suggestion
 *      |Callee saved registers | // multiple of 8 bytes, not including FP/RA
```

shushanhf replied:
thanks

shushanhf committed:
revert the OSR offset's changing which will be pushed by a new PR.

shushanhf commented:
I think the CI went wrong. e.g.
```
@azure-pipelines
runtime (Build Formatting linux x64) Failing after 4m — Build Formatting linux x64 failed
[Details](https://github.com/dotnet/runtime/pull/100962/checks?check_run_id=24011496326)
@azure-pipelines
runtime (Build Formatting windows x64) Failing after 7m — Build Formatting windows x64 failed
```
but format-patch is empty.

jakobbotsch commented:
> I think the CI went wrong. e.g.
> 
> ```
> @azure-pipelines
> runtime (Build Formatting linux x64) Failing after 4m — Build Formatting linux x64 failed
> [Details](https://github.com/dotnet/runtime/pull/100962/checks?check_run_id=24011496326)
> @azure-pipelines
> runtime (Build Formatting windows x64) Failing after 7m — Build Formatting windows x64 failed
> ```
> 
> but format-patch is empty.

You can run jit-format locally, see the docs at https://github.com/dotnet/jitutils and https://github.com/dotnet/jitutils/blob/main/doc/formatting.md.

am11 commented:
The underlying error shows up on Windows leg:

> D:\a\_work\1\s\venv\Scripts\python.exe D:\a\_work\1\s/src/coreclr/scripts/jitformat.py -r D:\a\_work\1\s -o windows -a x64
> ...
> 2024-04-19T09:58:48.1997764Z   No cdac-build-tool set or does not exist

started happening after https://github.com/dotnet/runtime/pull/100650 merge. cc @lambdageek

jakobbotsch reviewed (approved):
This change looks good to me now. Thanks for addressing all the feedback! I will give RISC-V folks some time to review as well.

tomeksowi reviewed (approved):
LGTM

@sirntar, you did more work with stack frames/OSR, could you review when you return on Monday?

sirntar commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -7851,80 +7790,57 @@ void CodeGen::genPushCalleeSavedRegisters(regNumber initReg, bool* pInitRegZeroe
+        // save callee-saved registers before using 'initReg' for the first time. Instead, we can use REG_SCRATCH
+        // beforehand. We don't care if REG_SCRATCH will be overwritten, so we'll skip 'RegZeroed check'.
+        // TODO-RV64: this should be resolved before calling `genPushCalleeSavedRegisters`.
+        genStackPointerAdjustment(-totalFrameSize, REG_SCRATCH, pInitRegZeroed, /* reportUnwindData */ true);
```
I don't think `pInitRegZeroed` should be changed here, because we are using `REG_SCRATCH`, not `initReg`, so we are changing some trash reg.
That said, this line should look like this:
```c++
genStackPointerAdjustment(-totalFrameSize, REG_SCRATCH, nullptr, /* reportUnwindData */ true);
```

On the arm in a similar case we use an additional variable to ignore the returned value, but imo we do not need it:
https://github.com/dotnet/runtime/blob/3092323b0f960e7a4c763049203c2cafe5b9b222/src/coreclr/jit/codegenarmarch.cpp#L4897-L4899

And here were the related RV64 PR, where similar lines were added: https://github.com/dotnet/runtime/pull/99353

shushanhf replied:
ok, thanks

sirntar commented on src/coreclr/jit/codegenriscv64.cpp:
```diff
@@ -7935,18 +7851,9 @@ void CodeGen::genPushCalleeSavedRegisters(regNumber initReg, bool* pInitRegZeroe
-        assert(frameType == 2);
-        calleeSaveSPDelta = totalFrameSize - calleeSaveSPDelta;
-        genStackPointerAdjustment(-calleeSaveSPDelta, initReg, pInitRegZeroed, /* reportUnwindData */ true);
+        genStackPointerAdjustment(-leftFrameSize, REG_SCRATCH, pInitRegZeroed, /* reportUnwindData */ true);
```
Similar case to previous one - imo should be changed to:
```c++
genStackPointerAdjustment(-leftFrameSize, REG_SCRATCH, nullptr, /* reportUnwindData */ true);
```

shushanhf replied:
ok, thanks

shushanhf committed:
amend codegenriscv64.cpp for CRs.

sirntar reviewed (approved):
LGTM

jakobbotsch commented:
@shushanhf Can you please avoid amending/force pushing to PRs? It makes it hard for us to see exactly what changed and to review the new changes. If you can push them as new commits that makes it much easier. We squash every PR on merge, so the history will be cleaned up anyway.

shushanhf commented:
> @shushanhf Can you please avoid amending/force pushing to PRs?

Ok, thanks, I will.

> It makes it hard for us to see exactly what changed and to review the new changes. If you can push them as new commits that makes it much easier. We squash every PR on merge, so the history will be cleaned up anyway.

Sorry, I didn't know this.
I thought the github had added a new button `Update branch` which I thought I had to update the PR's base as soon as possible.
And I thought the github reviewing is liking the gerrit which different patches within the same PR are easy to diff where independed of the base.
Maybe the github should add the similar feature.

jakobbotsch commented:
Thanks for helping me understand the motivation for the change, and for making LA64/RV64 more similar to ARM64 around OSR.

jakobbotsch commented:
@BruceForstall If you have any feedback around this then please feel free to leave it. To sum up the motivation of the change you can see https://github.com/dotnet/runtime/pull/100962#issuecomment-2058831736.
